README.md
# news-classification



Sample Usage : 


```terminal
Go to news-classification directory 
>> news.py <category> <news*i>
#Sample
>> news.py ekonomi 5
you can get economy news like that.
```


<h3>If you have not Mozilla Firefox browser</h3>
First of all you have to download webdriver for your browser. <br>
For Google Chrome <a href="https://sites.google.com/a/chromium.org/chromedriver/downloads">Chrome webdriver</a><br>
For Opera <a href="https://github.com/operasoftware/operachromiumdriver/releases">Opera webdriver</a><br>
For any other browser you can easily find webdriver if you search at google. <br>
After you downloaded your webdriver, move your webdriver's exe file to project directory. Then open news_parser.py file and change start_connection method's driver_path parameter to your exe file name and same methods body you have to change webdriver.Firefox to webdriver.YourBrowser(Chrome,Opera etc.). Now all is set enjoy it !.

news.txt
Dünya Borsalar Federasyonu (WFE) verilerine göre, söz konusu artışla dünya borsalarının değeri, ilk yarıda 82.3 trilyon dolara ulaşırken, New York Borsası sahip olduğu 24.2 trilyon dolarlık piyasa büyüklüğü ile zirvedeki yerini korudu. İlk yarıda piyasa değeri açısından en hızlı artış yaşanan borsa yüzde 60’la İran olurken, en fazla kayıp yüzde 11.5 ile Ukrayna borsasında görüldü.

ABD ile Çin arasında ticaret müzakerelerine dair haber akışının gündemden düşmediği ilk yarıda, korumacılık eğilimi endişeleri, Brexit belirsizliği, küresel büyüme hızının yavaşlaması, dünya çapında ekonomik göstergelerin zayıflaması ve jeopolitik riskler piyasalara yön veren gelişmeler olarak öne çıktı.
Tüm bunlar sonucunda yatırımcılar riskli varlıklardan kaçarken, tahvil piyasalarında getiri eğrilerinin terse döndüğü görüldü. Merkez bankaları tarafında ise ABD Merkez Bankası (Fed) ile Avrupa Merkez Bankası (ECB) başta olmak üzere, birçok gelişmiş ve gelişmekte olan ülkede faizlerde düşüş sinyali verilerek genişlemeci politikalara geri dönüşün ilk adımları atıldı.

model.py
from train import randomforest_predict,model_columns
import prepare_news as pn
import pandas
import sys

# sys.argv[1] file name

f = open(sys.argv[1])
_news_from_file = f.read()
news = pn.conjuction_prepositions(pn.listModel(pn.clear(_news_from_file)))

#prepare news
col = model_columns()
data = {}
for i in col:
    data[i] = [0]

for j in news:
    for k in col:
        if k in j:
            data[k][0]+=1
summ = 0
for i in col:
    summ += data[i][0]

for i in col:
    if summ!=0:
        data[i][0] /= summ
#prepare news
data = pandas.DataFrame(data)


category = randomforest_predict(data)

print("""
*************************************** Our Models Prediction ************************************************\n""")
print(category)


print("""
*************************************** Actual Category ************************************************\n""")
print("Ekonomi")

terms.py
from bs4 import BeautifulSoup
import requests


def routineSoupPreparation(url):
    #Routine things
    response = requests.get(url)

    soup = BeautifulSoup(response.text,'html.parser')
    return soup
    


class Terms():
    #Scraping terms need's to clear. Not fits the news style.
    __terms = ""
    def economy(self):
        soup = routineSoupPreparation('https://www.uyumsoft.com/bilmeniz-gereken-temel-ekonomi-terimleri/')
        terms = soup.find('div',{'class':'avia_textblock av_inherit_color'}).find_all('strong')
        self.__terms += " " + str(terms)
    def politics(self):
        soup = routineSoupPreparation('https://www.on5yirmi5.com/haber/guncel/politika/462/siyasi-terimler-sozlugu.html')
        terms = soup.find('div',{'class':'article-content article-body-responsive col-lg-9'}).find_all('strong')
        self.__terms += " " + str(terms)
    def world(self):
        #find better website for scraping...
        soup = routineSoupPreparation('https://www.dunyaatlasi.com/cografya-terimleri-sozlugu/')
        terms = soup.find('div',{'id':'searchContent'}).find_all('strong')
        self.__terms += " " + str(terms)
    def sport(self):
        soup = routineSoupPreparation('https://www.dersimiz.com/terimler-sozlugu/Futbol-Terimleri-Sozlugu-70-Sayfa-2.html')
        terms = soup.find('article').find_all('strong')
        self.__terms += " " + str(terms)
    def health(self):
        soup = routineSoupPreparation('https://blog.konusarakogren.com/ingilizce-tibbi-terimler/')
        
        terms = soup.find('div',{'class','post-inner'}).find_all('p')
        self.__terms += " " + str(terms)
    
    def terms(self):
        return self.__terms.lower()

    def collectTerms(self):
        self.economy()
        self.politics()
        #self.world()
        self.sport()
        self.health()

def getTermsList():
    """
    get all terms as a string data type. 
    and clear all of the useless_tags from this terms string
    after clearization return list of these terms.
    """
    t = Terms()
    t.collectTerms()
    _terms = t.terms() #terms are set and sets are not iterable
    useless_tags = ['strong','<br/>','<br>','<p','p>','\r','\n','>','<','/','!','[',']',':',',','\xa0']
        
    for i in useless_tags:
        _terms =_terms.replace(i,' ')
    
    return list(set(_terms.split()))


train.py
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm


data = pd.read_csv('data.csv')

# find useless attributes on dataset
useless_attributes = [col for col in data.columns if data[col].std()==0]
# when you find the useless attributes drop all columns
data.drop(useless_attributes,axis=1,inplace=True)

# create categories


data['Category'] = "none"

# <Ekonomi> News : 220
# <Spor> News : 220
# <Politika> News : 240
# <Yasam> News : 220
# <Dunya> News : 200

data.loc[0:220,'Category'] = "ekonomi"
data.loc[220:440,'Category'] = "spor"
data.loc[440:680,'Category'] = "politika"
data.loc[680:900,'Category'] = "yasam"
data.loc[900:1100,'Category'] = "dunya"

# continue with clean data
clear_data = data.loc[:899]


# cross validation
train_x,test_x,train_y,test_y = train_test_split(clear_data.drop('Category',axis=1),clear_data.Category,test_size =0.2)

models = {'K-nn':KNeighborsClassifier(n_neighbors = 15),
'RandomForest':RandomForestClassifier(),
'DecisionTree':DecisionTreeClassifier(),
'SVM':svm.SVC(gamma = 1,C=0.8)}
# create knn object
# knn = KNeighborsClassifier(n_neighbors = 15)
# knn.fit(train_x,train_y)

#fit data
for i in models.values():
    i.fit(train_x,train_y)

#find models scores
for i,j in zip(models.keys(),models.values()):
    print(i,"guesses :",j.score(test_x,test_y))

def randomforest_predict(news):
    global models
    return models['RandomForest'].predict(news)

def model_columns():
    global clear_data
    return clear_data.drop('Category',axis=1).columns

# predict test data


# model_predicts = knn.predict(test_x)
# model_predicts = models['knn'].predict(test_x)
# test_y = np.array(test_y)

# true_guess = 0
# for i in range(len(model_predicts)):
#     if model_predicts[i] == test_y[i]:
#         true_guess+=1

# print("True guesses :",true_guess)
# print('Model accuracy :',models['RandomForest'].score(test_x,test_y))



news_parser.py
from selenium import webdriver
#from selenium.webdriver.common.keys import Keys
import time

driver = None

def start_connection(driver_path="geckodriver.exe",url = 'https://tr.sputniknews.com/'):
    #__url = url +sys.argv[1]+'/' #https://tr.sputniknews.com/ekonomi/
    global driver
    driver = webdriver.Firefox(executable_path = driver_path)
    driver.get(url)



def get_links(click_time = 1):
    #time.sleep(1)
    print('Searching....')

    for i in range(click_time):
        time.sleep(1.2)
        driver.find_element_by_css_selector('.b-btn.m-more').click()



    selen = driver.find_elements_by_css_selector('.b-stories__img')
    print('News length :',len(selen))
    links = []
    for i in selen:
        links.append(i.get_attribute('href'))


    return links

def close_connection():
    driver.close()





prepare_news.py
#import news #You can get news from there.
import string
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')
from collections import Counter
import pandas as pd
#or any excell file doesn't matter.

"""
we have news that we can read but our machine learning model can't. Because of this we
have to prepare news for our machine learning model. So how can we do
this, first our news should a list that contains only words. Let me give
you an example:
    news_content = ["This summer python programming language is going to rise."]
    first of all we dont want any dots(.) .
    news_content = ["This summer python programming language is going to rise"]
    then let's do what we wanted first
    news_content = ['This','summer','python','programming','language','is','going','to','rise']
    conjuctions and prepositions worthless for our model.
    so lets clear our model one more step
    news_content = ['summer','python','programming','language','rise']
    Data Model:
    Id summer python programming language rise
    0  20%    20%    20%          20%     20%
    or better way
    Id summer python programming language rise
    0  0.5    0.5    0.5          0.5     0.5
    When we get a data model like this. We can train our machine learning model
"""
__conjuctions_prepositions_TR = [] #constant conjuctions and prepositions for removal
__model_columns = [] #Means words that model contains
#our training data has to give us the information about model columns.


def clear(content):
    #we should clear dummy characters from content
    dummy_characters = ['©','\n','\xa0','\t']
    for i in dummy_characters:
        content = content.replace(i,'')
    
    content = content.translate(str.maketrans('','',string.punctuation))
    return content

def listModel(content):
    content =  content.split(' ')
    while '' in content:
        content.remove('')
    while '\n' in content:
        content.remove('\n')
    content = [x.lower() for x in content]

    return content

__conjuctions_prepositions_TR = stopwords.words('turkish')

def conjuction_prepositions(word_list):
    #Step 3
    for i in __conjuctions_prepositions_TR:
        for word in word_list:
            if word == i:
                word_list.remove(i)

    return word_list

def create_model(list_words):

    unique_word_list =  list(Counter(list_words).keys())
    counter_of_list = list(Counter(list_words).values())
    df = pd.DataFrame(counter_of_list)
    df = df.transpose()
    df.columns = unique_word_list
    df =  df.sort_values(by = 0, axis=1, ascending=False)
    return df

def frequency(dataframe):
    updated_dataframe = dataframe.apply(lambda x: x / int(dataframe.sum(axis=1)), axis=1)
    return  updated_dataframe

def final_cleaning(final_data):
    word_list = ['bir','iki','üç','dört','beş','altı','yedi','sekiz','dokuz','sıfır']
    for i in word_list:
        for item in final_data:
            if item == i:
                final_data.remove(i)
    for item in final_data:
        item = item.replace("'",'')
        item = item.replace('nin','')
        item = item.replace('ni','')

    return final_data




news.py
import requests
from bs4 import BeautifulSoup
import sys
import time
import news_parser as parser
import pandas
import prepare_news as pn
import terms


class News():
    def __init__(self,category,title,content):
        self.__category = category
        self.__title = title
        self.__content = content

    

    def getCategory(self):
        return self.__category
    def getTitle(self):
        return self.__title
    def getContent(self):
        return self.__content
    def __str__(self):
        return "Category<{0}>\nTitle<{1}>\nContent<{2}>".format(self.__category,self.__title,self.__content)

# __url = "https://tr.sputniknews.com/"+sys.argv[1]+"/"
# parser.start_connection(url =__url )
# news_link = None
# #if news size parameter entered.
# if len(sys.argv)>2:
#     news_link = parser.get_links(int(sys.argv[2]))
# else:
#     news_link = parser.get_links()

# parser.close_connection()

def clean(dirty_content):
    return dirty_content.text #return clean content

def _news_content(url,category):
    #get response from url 
    _response = requests.get(url) # if 200 ok.

    _soup = BeautifulSoup(_response.text,'html.parser') #create html parser
    _dirty_content = _soup.find('div',{'class':'b-article__text'}) #get dirty_content

    _clean_content = clean(_dirty_content)

    #we need to get title too.
    return News(category,'title',_clean_content)


def __news_objects(news_link,category):
    news = []
    print("Searching......")
    for i in news_link:
        #time.sleep(0.5)
        news.append(_news_content(i,category))

    return news

# def _clear_content(content):
#     #© this sign exists
#     _dirty = str(content)
#     _dirty = _dirty.replace('©','')
#     _dirty = _dirty.replace('\n','')
#     return _dirty




# news_objects = __news_objects(news_link,sys.argv[1])
# news_data = [[]]
# last_news = None
# for i in news_objects:
#     news_data.append([i.getCategory(),i.getTitle(),_clear_content(i.getContent())])
#     last_news = _clear_content(i.getContent())



"""
NOTE !!!!
this code block written for only test purposes on the real usage
you can't use this code because, values are not lists here and cout variable 
initalizes to 0 for once. If you solve this issues you can use this code.
"""

news_terms = terms.getTermsList() #Our dataframes columns
data = {}
#initialize dictionarys columns
# for i in news_terms:
#     data[i] =[0 for i in range(60)] 

#dataframe key,value
#data['key'] = value..  // data['key'] = [values...]

#after initialization of data columns now fill one row with one news
#means you will calculate the frequency of each terms
# for i in range(len(news_objects)):
#     cout = 0
#     news = pn.conjuction_prepositions(pn.listModel(pn.clear(news_objects[i].getContent())))
#     for j in news:
#         for k in news_terms:
#             if k in j:
#                 cout +=1 #cout storing for % 
#                 data[k][i]+=1


# #this loops for finding repeating percents. %
# summ = 0
# for i in news_terms:
#     for j in range(len(news_objects)):
#         summ += data[i][j]
#     for j in range(len(news_objects)):
#         if summ!=0:
#             data[i][j] /= summ
#     summ = 0 

def help_clean(data):
    return pn.conjuction_prepositions(pn.listModel(pn.clear(data)))


def getMultipleNews(category):
    global news_terms
    global data
    # get so much link
    __url = "https://tr.sputniknews.com/"+"ekonomi"+"/"
    parser.start_connection(url =__url)
    links = parser.get_links(13)
    parser.close_connection()
    __url = "https://tr.sputniknews.com/"+"spor"+"/"
    parser.start_connection(url =__url)
    links += parser.get_links(13)
    parser.close_connection()
    __url = "https://tr.sputniknews.com/"+"politika"+"/"
    parser.start_connection(url =__url)
    links += parser.get_links(13)
    parser.close_connection()
    __url = "https://tr.sputniknews.com/"+"yasam"+"/"
    parser.start_connection(url =__url)
    links += parser.get_links(13)
    parser.close_connection()
    __url = "https://tr.sputniknews.com/"+"dunya"+"/"
    parser.start_connection(url =__url)
    links += parser.get_links(13)
    parser.close_connection()

    for i in news_terms:
        data[i] =[0 for i in range(len(links))] 

    news = __news_objects(links,category)
    for i in range(len(news)):
        cout = 0
        __news = pn.conjuction_prepositions(pn.listModel(pn.clear(news[i].getContent())))
        for j in __news:
            for k in news_terms:
                if k in j:
                    cout +=1 #cout storing for % 
                    data[k][i]+=1

    #this loops for finding repeating percents. %
    summ = 0
    for i in news_terms:
        for j in range(len(news)):
            summ += data[i][j]
        for j in range(len(news)):
            if summ!=0:
                data[i][j] /= summ
        summ = 0 


"""
this method should work for calculating % data[i] /= cout 
not working on this version i didn't debug it .

#calculate the percent of repeating number
for i in news_terms:
    try:
        data[i] /= cout
    except:
        #for zero division error
        pass
"""

getMultipleNews('ekonomi')
# getMultipleNews('spor')
# getMultipleNews('politika')
# getMultipleNews('saglik')

print(data) #Sample data output.
dataFrame = pandas.DataFrame(data)
print(dataFrame)
dataFrame.to_csv('data.csv',index=False)

#print dataFrame


# list_of_words = []
# for i in news_objects:
#     list_of_words = list_of_words+pn.conjuction_prepositions(pn.listModel(pn.clear(i.getContent())))
#     print(pn.conjuction_prepositions(pn.listModel(pn.clear(i.getContent())))) #what i want
    


# model = pn.frequency(pn.create_model(pn.final_cleaning(list_of_words)))

# model.to_csv("output.csv")


